ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsReader.scala

```diff
@@ -126,36 +138,44 @@ private[dao] final class TransactionsReader(
       requestingParties: Set[Party],
       verbose: Boolean,
   ): Source[(Offset, GetTransactionTreesResponse), NotUsed] = {
-    val events: Source[EventsTable.Entry[TreeEvent], NotUsed] =
-      PaginatingAsyncStream.streamFrom((startExclusive, Option.empty[Int]), getOffset[TreeEvent]) {
-        case (prevOffset, prevNodeIndex) =>
-          val query =
-            EventsTable
-              .preparePagedGetTransactionTrees(sqlFunctions)(
-                startExclusive = prevOffset,
-                endInclusive = endInclusive,
-                requestingParties = requestingParties,
-                pageSize = pageSize,
-                previousEventNodeIndex = prevNodeIndex
-              )
-              .withFetchSize(Some(pageSize))
-          val rawEvents =
-            dispatcher.executeSql(dbMetrics.getTransactionTrees) { implicit connection =>
-              query.asVectorOf(EventsTable.rawTreeEventParser)
-            }
-          rawEvents.flatMap(
-            es =>
-              Timed.future(
-                future = Future.traverse(es)(deserializeEntry(verbose)),
-                timer = dbMetrics.getTransactionTrees.translationTimer,
+
+    val eventsRangeF: Future[EventsRange[Long]] = dispatcher.executeSql(dbMetrics.getRowIdRange) {
+      connection =>
+        EventsRange.eventsRange(EventsRange(startExclusive, endInclusive))(connection)
+    }
+
+    def getEvents(eventsRange0: EventsRange[Long]): Source[EventsTable.Entry[TreeEvent], NotUsed] =
```

Comment:
Could `eventsRangeForFirstPage` be a more telling name?

Reply:
it is NOT for the first page only. This function is used to extracts `row_ids`/`EventsRange[Long]` from the passed `TreeEvent`/`Event`, it is used for the pagination.